crates/musli-zerocopy/src/pointer/ref.rs

```diff
@@ -156,6 +192,25 @@ where
     E: ByteOrder,
     O: Size,
 {
+    fn from_parts(offset: O, metadata: T::Stored<O>) -> Self {
+        match Self::try_from_parts(offset, metadata) {
+            Ok(ok) => ok,
+            Err(_) => panic!("Metadata describes an invalid layout"),
+        }
+    }
+
+    fn try_from_parts(offset: O, metadata: T::Stored<O>) -> Result<Self, Error> {
+        if <T as Pointee>::pointee_layout::<E, O>(metadata.swap_bytes::<E>()).is_some() {
+            Ok(Self {
+                offset,
+                metadata,
+                _marker: PhantomData,
+            })
+        } else {
+            Err(Error::new(ErrorKind::InvalidLayout))
+        }
+    }
```

Comment:
I think we need to check for `usize` overflow when constructing the ref, something like this?

```suggestion
    #[inline]
    fn from_parts(offset: O, metadata: T::Stored<O>) -> Self {
        match Self::try_from_parts(offset, metadata) {
            Ok(ok) => ok,
            Err(error) => panic!("{error}"),
        }
    }

    #[inline]
    fn try_from_parts(offset: O, metadata: T::Stored<O>) -> Result<Self, Error> {
        let Ok(layout) = T::pointee_layout::<E, O>(metadata.swap_bytes::<E>()) else {
            return Err(Error::new(ErrorKind::InvalidLayout));
        };

        let o = offset.as_usize::<E>();

        if o.checked_add(layout.size()).is_none() {
            return Err(Error::new(ErrorKind::InvalidOffsetRange {
                offset: usize::into_repr(o),
                max: usize::into_repr(usize::MAX - layout.size()),
            }));
        };

        Ok(Self {
            offset,
            metadata,
            _marker: PhantomData,
        })
    }
```